keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1475,225 +1382,400 @@ prevented. Deleting the entire cluster also leaves resources allocated outside
+    //
+    // +optional
+    RequestNames []string
 
-    // VendorParameters are arbitrary setup parameters for all requests of the
-    // claim. They are ignored while allocating the claim.
-    VendorParameters runtime.Object
+    // If set, then the devices must have this attribute and its value must be the same.
+    //
+    // For example, if you specified "numa.dra.example.com" (a hypothetical example!),
+    // then only devices in the same NUMA node will be chosen.
+    //
+    // +optional
+    // +listType=atomic
+    MatchAttribute *string
 
-    // Requests describes all resources that are needed from the driver.
-    Requests []ResourceRequest
+    // Future extension, not part of the current design:
+    // A CEL expression which compares different devices and returns
+    // true if they match.
+    //
+    // Because it would be part of a one-of, old schedulers will not
+    // accidentally ignore this additional, for them unknown match
+    // criteria.
+    //
+    // matcher string
 }
 
-// ResourceRequest is a request for resources from one particular driver.
-type ResourceRequest struct {
-    // VendorParameters are arbitrary setup parameters for the requested
-    // resource. They are ignored while allocating a claim.
-    VendorParameters runtime.Object
+type ClaimConfigurationParameters struct {
+    // The configuration applies to devices in these requests.
+    //
+    // If empty, the configuration applies to all devices in the claim.
+    //
+    // +optional
+    RequestNames []string
 
-    ResourceRequestModel
+    ConfigurationParameters // inline
 }
 
-// ResourceRequestModel must have one and only one field set.
-type ResourceRequestModel struct {
-    // NamedResources describes a request for resources with the named resources model.
-    NamedResources *NamedResourcesRequest
+// ConfigurationParameters must have one and only one field set. It gets embedded
+// inline in some other structs which have other fields, so field names must
+// not conflict with those.
+type ConfigurationParameters struct {
+    Opaque *OpaqueConfigurationParameters
 }
 
-// NamedResourcesRequest is used in ResourceRequestModel.
-type NamedResourcesRequest struct {
-    // Selector is a CEL expression which must evaluate to true if a
-    // resource instance is suitable. The language is as defined in
-    // https://kubernetes.io/docs/reference/using-api/cel/
+// OpaqueConfigurationParameters contains configuration parameters for a driver.
+type OpaqueConfigurationParameters struct {
+    // DriverName is used to determine which kubelet plugin needs
+    // to be passed these configuration parameters.
+    //
+    // An admission webhook provided by the driver developer could use this
+    // to decide whether it needs to validate them.
+    //
+    // Must be a DNS subdomain and should end with a DNS domain owned by the
+    // vendor of the driver.
+    DriverName string
```

Comment:
If this moved up to `type ConfigurationParameters` (and was optional) then I think it would be simpler:

```
type ConfigurationParameters struct {
    DriverName string // optional

    // one-of
    Opaque *runtime.RawExtension
}
```

Reply:
https://github.com/kubernetes/enhancements/pull/4709#discussion_r1642685097

Reply:
The driver name for a runtime.RawExtension config cannot be optional, otherwise it is unclear without peeking into the content which driver they are for.

But looking at `ClaimConfiguration[Parameters]` I see that must have made a mistake during editing at some point: that should embed `OpaqueConfiguration[Parameters]` which includes the required driver name. I need to double-check that.

Reply:
If you don't specify a drivername, pass it to all of them?

Reply:
And then they have to decode the configuration of some other driver to determine whether it is "for them". That seems fragile and error prone: decoding errors would have to be treated as "must be for someone else" and ignored when it really was meant for the driver and just has a typo somewhere.

Reply:
Or it causes a decode error which causes the pod to fail, and the user should fix the YAML?  I guess that's pretty ugly, but needs to be possible anyway.  OK.